Docs: Add "Going to production" page for App Router (#59304)
vercel/next.js: docs/02-app/01-building-your-application/08-deploying/01-production-checklist.mdx

---
title: Going to Production
description: Before taking your Next.js application to production, here are some recommendations to ensure the best user experience.
---

Before taking your Next.js application to production, here are some recommendations to ensure the best user experience.

> ## In General

Reviewer: ```suggestion
## Overview
```

`@next/bundle-analyzer` is a plugin for Next.js that helps manage the size of your JavaScript files. It shows you how big each part of your webpack output is. This is helpful because you can find and fix parts of your code that are too large. For example, you might remove large dependencies, split your code, or only load some parts when needed.

How to use `@next/bundle-analyzer`:

1. **Install**: Run `npm install @next/bundle-analyzer webpack-bundle-analyzer`.
2. **Configure**: Add the bundle analyzer's settings to your `next.config.js`.

```js
const withBundleAnalyzer = require('@next/bundle-analyzer')({
  enabled: process.env.ANALYZE === 'true',
})
module.exports = withBundleAnalyzer({})
```

3. **Run Analysis**: Use `ANALYZE=true npm run build` to start looking at your file sizes.
4. **Review Findings**: Check the report to see where you can make improvements.

Doing this regularly while you develop and before you launch your site is important to keep your code efficient.

## Logging

Since Next.js runs on both the client and server, there are multiple forms of logging supported:

- `console.log` in the browser

> - `stdout` on the server

Reviewer: ```suggestion
- `process.stdout.write` on the server
```

If you want a structured logging package, we recommend [Pino](https://www.npmjs.com/package/pino). If you're using Vercel, there are [pre-built logging integrations](https://vercel.com/integrations?utm_source=next-site&utm_medium=docs&utm_campaign=next-website#logging) compatible with Next.js.

> ## **Error Handling**

Reviewer: ```suggestion
## Error Handling
```

> Proper error handling is key to keeping your application stable and providing a good user experience. This is especially important with the dynamic routes and server actions in the App Router.

Reviewer: ```suggestion
Proper error handling is key to keeping your application stable and providing a good user experience. This is especially important with the Dynamic Route Segments and Server Actions in the App Router.
```

**Use of `error.js` for Error Handling**:

> The `error.js` file convention is used for gracefully handling runtime errors in nested routes. It automatically wraps a route segment and its nested children in a React Error Boundary, allowing you to create a tailored error UI and isolate errors to specific segments while keeping the rest of the application functional.

Reviewer: ```suggestion
The [`error.js` file](/docs/app/building-your-application/routing/error-handling) convention is used for gracefully handling runtime errors in nested routes. It automatically wraps a route segment and its nested children in a React Error Boundary, allowing you to create a tailored error UI and isolate errors to specific segments while keeping the rest of the application functional.
```

> **Handling Server Errors**:

Reviewer: Maybe you could combine these two sections, since they are both referring to `error.js`

```suggestion
```

If an error is thrown inside a Server Component, Next.js forwards an `Error` object to the nearest `error.js` file. This object includes a generic message and a digest property in production as a security measure to prevent leaking sensitive details.

## Loading Performance

> To improve loading performance, you first need to determine what to measure and how to measure it. [Core Web Vitals](https://vercel.com/blog/core-web-vitals?utm_source=next-site&utm_medium=docs&utm_campaign=next-website) is a good industry standard that is measured using your own web browser. If you are not familiar with the metrics of Core Web Vitals, review this [blog post](https://vercel.com/blog/core-web-vitals?utm_source=next-site&utm_medium=docs&utm_campaign=next-website) and determine which specific metric/s will be your drivers for loading performance. Ideally, you would want to measure the loading performance in the following environments:

Reviewer: Suggestion for improving clarity on what core web vitals are, and removing the duplicate blog post link. 

```suggestion
To improve loading performance, you first need to determine what to measure and how to measure it.

[Core Web Vitals](https://vercel.com/blog/core-web-vitals?utm_source=next-site&utm_medium=docs&utm_campaign=next-website) are metrics created by Google to measure the end-user experience of web pages. These metrics are important as they influence Google's ranking. 
```

- In the lab, using your own computer or a simulator.
- In the field, using real-world data from actual visitors.
- Local, using a test that runs on your device.

> - Remote, using a test that runs in the cloud.

Reviewer: Since these are from the original /pages docs, how do you feel about cutting it? 

They are good to know, but unless we give user specific examples on how to test in these  environments, I don't think it's helpful information. 

Perhaps, it may be more helpful to show them how to measure Core Web Vitals using lighthouse, or Vercel's dashboard (if deployed to Vercel). 

From the overview: 

> Run [Lighthouse](https://developers.google.com/web/tools/lighthouse) to check for performance, best practices, accessibility, and SEO. For best results, use a production build of Next.js and use incognito in your browser so results aren't affected by extensions.

> Once you are able to measure the loading performance, use the following strategies to improve it iteratively so that you apply one strategy, measure the new performance and continue tweaking until you do not see much improvement. Then, you can move on to the next strategy.

Reviewer: Improving sentence clarity

```suggestion
Once you are able to measure the loading performance, use the following strategies to improve it iteratively. That is, if you apply one strategy, measure the new performance and continue tweaking until you do not see much improvement. Then, you can move on to the next strategy.
```

- **Optimize Caching**: Use caching strategies suited for dynamic content, as detailed in the caching section.
- **Server Actions Efficiency**: Optimize server actions for quick data retrieval and minimal processing.

> - **Minimize JavaScript Load**: Remove unused JavaScript and implement code splitting. Dynamically load components as needed.

Reviewer: ```suggestion
- **Reduce JavaScript Bundle Sizes**: Remove unused JavaScript, and move data fetching and logic to the server.
```

  - Setting up your Code Editor to view import costs and sizes
  - Finding alternative smaller packages
  - Dynamically loading components and dependencies

> - **Parallel Route Rendering**: Leverage the App Router’s ability to render multiple components in parallel for enhanced performance.

Reviewer: ```suggestion
- **Streaming**: Leverage the Loading UI and Suspense to stream components in parallel for improved initial load.
```

---
title: Going to Production
description: Before taking your Next.js application to production, here are some recommendations to ensure the best user experience.
---

Before taking your Next.js application to production, here are some recommendations to ensure the best user experience.

## Overview

- Use [caching](/docs/app/building-your-application/caching) wherever possible.
- Utilize logging for both the client and server.
- Improve [loading performance](#loading-performance)

> - Leverage [Server Components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) as much as possible for improved performance. Unlike [Client Components](https://nextjs.org/docs/app/building-your-application/rendering/client-components) which require JavaScript on the client, Server Components render on the server, which significantly reduces the size of your client-side JavaScript bundle. See [Composition Patterns](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns) for best practices on using Server and Client Components.

Reviewer 2: ```suggestion
- Leverage [Server Components](/docs/app/building-your-application/rendering/server-components) as much as possible for improved performance. Unlike [Client Components](/docs/app/building-your-application/rendering/client-components) which require JavaScript on the client, Server Components render on the server, which significantly reduces the size of your client-side JavaScript bundle. See [Composition Patterns](/docs/app/building-your-application/rendering/composition-patterns) for best practices on using Server and Client Components.
```

- Ensure your database and backend are deployed in the same region.
- Prevent sensitive data from being exposed to the client with [tainting](/docs/app/building-your-application/data-fetching/patterns#preventing-sensitive-data-from-being-exposed-to-the-client)

> - Ensure errors in nested route segments are handled gracefully with the [error.js file](https://nextjs.org/docs/app/building-your-application/routing/error-handling)

Reviewer 2: ```suggestion
- Ensure errors in nested route segments are handled gracefully with the [error.js file](/docs/app/building-your-application/routing/error-handling)
```

> - Customize your 404 (Not Found) responses using the [`notFound()`](/docs/app/api-reference/functions/not-found) function and [`not-found.js` file](https://nextjs.org/docs/app/api-reference/file-conventions/not-found)

Reviewer 2: ```suggestion
- Customize your 404 (Not Found) responses using the [`notFound()`](/docs/app/api-reference/functions/not-found) function and [`not-found.js` file](/docs/app/api-reference/file-conventions/not-found)
```

> - During the development phase, run [Lighthouse](https://developers.google.com/web/tools/lighthouse) to check for performance, best practices, accessibility, and SEO. For best results, use a production build of Next.js and use incognito in your browser so results aren't affected by browser extensions.

Reviewer 2: ```suggestion
- During the development phase, run [Lighthouse](https://developers.google.com/web/tools/lighthouse) to check performance, best practices, accessibility, and SEO. For best results, use a production build of Next.js and use incognito in your browser so results aren't affected by browser extensions.
```

> - Once in production, ensure you are measuring performance and collecting real user metrics (RUM) with [analytics](/docs/app/building-your-application/optimizing/analytics) to make sure you don't regress.

Reviewer 2: ```suggestion
- Once in production, ensure you are measuring performance and collecting real user metrics (RUM) with [analytics](/docs/app/building-your-application/optimizing/analytics).
```

- Review [Supported Browsers and Features](/docs/architecture/supported-browsers)
- Optimize the user experience with built-in components:
  - [Images](/docs/app/building-your-application/optimizing/images)
  - [Fonts](/docs/app/building-your-application/optimizing/fonts)
  - [Scripts](/docs/app/building-your-application/optimizing/scripts)
  - [Link](/docs/app/building-your-application/routing/linking-and-navigating)

> - Follow the recommended [security best practices for Server Actions (https://nextjs.org/blog/security-nextjs-server-components-actions).

Reviewer 2: ```suggestion
- Follow the recommended [security best practices for Server Actions (/blog/security-nextjs-server-components-actions).
```

- Consider adding a [Content Security Policy](/docs/app/building-your-application/configuring/content-security-policy)

## Caching

> Caching is essential for improving response times and reducing requests to backend services. The App Router in Next.js offers a robust [caching mechanism](https://nextjs.org/docs/app/building-your-application/caching) that allows to leverage static and dynamic rendering, as well as [revalidation](/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#revalidating-data).

Reviewer 2: ```suggestion
Caching is essential for improving response times and reducing requests to backend services. The App Router in Next.js offers a robust [caching mechanism](/docs/app/building-your-application/caching) that allows you to leverage static and dynamic rendering, as well as [revalidation](/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#revalidating-data).
```

### Static assets

Next.js automatically adds caching headers to immutable assets in **`/public`** or other static directories. Assets like CSS, static images, and other media are ideal candidates for long-term caching as they rarely change. The default caching headers applied are:

```jsx
Cache-Control: public, max-age=31536000, immutable
```

> This setup ensures that these assets are cached efficiently in the browser, reducing load times for returning users.

Reviewer 2: ```suggestion
This ensures that these assets are cached efficiently in the browser, reducing load times for returning users.
```

### Data caching

Next.js's built-in [Data Cache](/docs/app/building-your-application/caching#data-cache) persists results of data fetches across server requests and deployments. By default, fetch requests are cached, allowing data to be reused in subsequent requests. This feature is particularly useful for content that doesn't change frequently.

The caching behavior of data requests can be controlled through various strategies. Next.js supports [time-based revalidation](/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#time-based-revalidation) (automatically refreshing the cache after a set period) and [on-demand revalidation](/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#on-demand-revalidation) (refreshing the cache based on specific events).

### Controlling the cache/caching with Server Actions

> While Server Actions in Next.js are primarily used for mutations (like POST, PUT, DELETE operations), they also play a crucial role in cache management. After a mutation, it's crucial to ensure that the cached data reflects the latest state. Next.js facilitates this by providing APIs such as [`revalidatePath`](https://nextjs.org/docs/app/api-reference/functions/revalidatePath) and [`revalidateTag`](https://nextjs.org/docs/app/api-reference/functions/revalidateTag). These APIs can be used to purge or revalidate the cache immediately after a mutation. This capability ensures that any changes made by Server Actions are promptly reflected in the cached data, maintaining data consistency and accuracy.

Reviewer 2: ```suggestion
While Server Actions in Next.js are primarily used for mutations (like POST, PUT, DELETE operations), they also play a crucial role in cache management. After a mutation, it's crucial to ensure that the cached data reflects the latest state. Next.js facilitates this by providing APIs such as [`revalidatePath`](/docs/app/api-reference/functions/revalidatePath) and [`revalidateTag`](/docs/app/api-reference/functions/revalidateTag). These APIs can be used to purge or revalidate the cache immediately after a mutation. This capability ensures that any changes made by Server Actions are promptly reflected in the cached data, maintaining data consistency and accuracy.
```

> ### Opting Out of Caching

Reviewer 2: ```suggestion
### Opting out of caching
```

## Logging

Since Next.js runs on both the client and server, there are multiple forms of logging supported:

- `console.log`: Works in both browser and server. In the browser, it logs to the console. On the server, it outputs to the terminal.
  If you want a structured logging package, we recommend [Pino](https://www.npmjs.com/package/pino). If you're using Vercel, there are [pre-built logging integrations](https://vercel.com/integrations?utm_source=next-site&utm_medium=docs&utm_campaign=next-website#logging) compatible with Next.js.

## Error Handling

Proper error handling is key to keeping your application stable and providing a good user experience. This is especially important with the Dynamic Route Segments and Server Actions in the App Router, as they introduce complexities like variable user requests and potential mismatches with predefined routes, requiring handling to prevent errors and ensure smooth navigation.

> **Use of `error.js` for Error Handling**:

Reviewer 2: ```suggestion
### Using `error.js` for error handling
```

The [`error.js` file](/docs/app/building-your-application/routing/error-handling) convention is used for gracefully handling runtime errors in nested routes. It automatically wraps a route segment and its nested children in a React Error Boundary, allowing you to create a tailored error UI and isolate errors to specific segments while keeping the rest of the application functional.

If an error is thrown inside a Server Component, Next.js forwards an `Error` object to the nearest `error.js` file. This object includes a generic message and a digest property in production as a security measure to prevent leaking sensitive details.

> ## Loading Performance

Reviewer 2: ```suggestion
## Loading performance
```

- [Import Cost](https://marketplace.visualstudio.com/items?itemName=wix.vscode-import-cost) – Shows the size of each package you import right in VSCode.
- [Package Phobia](https://packagephobia.com/) – Helps you see the size impact of adding new dev dependencies to your project.
- [Bundle Phobia](https://bundlephobia.com/) - Tells you how much a new dependency could increase your JavaScript size.
- [Webpack Bundle Analyzer](https://github.com/vercel/next.js/tree/canary/packages/next-bundle-analyzer) – Creates a visual, interactive map of your webpack output files to show where space is used.
- [bundlejs](https://bundlejs.com/) - A browser tool for bundling and minifying projects, showing compressed sizes.

Additionally, make sure to implement the recommended optimization strategies.

- Minimize Client-Side JavaScript - It's important to keep JavaScript on the client-side as small as possible for quicker website loading and a better user experience.

> - Leverage [Server Components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) for efficient rendering. Server Components enable rendering on the server, which can significantly reduce the size of your client-side JavaScript bundle.

Reviewer 2: ```suggestion
- Leverage [Server Components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) for efficient rendering – Server Components enable rendering on the server, which can significantly reduce the size of your client-side JavaScript bundle.
```

> - Fetching data on the server: https://nextjs.org/docs/app/building-your-application/data-fetching/patterns#fetching-data-on-the-server

Reviewer 2: ```suggestion
- [Fetching data on the server](https://nextjs.org/docs/app/building-your-application/data-fetching/patterns#fetching-data-on-the-server)
```

#### Optimizing with `@next/bundle-analyzer`

[`@next/bundle-analyzer`](https://www.npmjs.com/package/@next/bundle-analyzer) is a plugin for Next.js that helps manage the size of your JavaScript files. It shows you how big each part of your webpack output is. This is helpful because you can find and fix parts of your code that are too large. For example, you might remove large dependencies, split your code, or only load some parts when needed.

How to use `@next/bundle-analyzer`:

1. **Install**: Run `npm install @next/bundle-analyzer webpack-bundle-analyzer`.
2. **Configure**: Add the bundle analyzer's settings to your `next.config.js`.

```js
const withBundleAnalyzer = require('@next/bundle-analyzer')({
  enabled: process.env.ANALYZE === 'true',
})
module.exports = withBundleAnalyzer({})
```

3. **Run Analysis**: Use `ANALYZE=true npm run build` to start looking at your file sizes.
4. **Review Findings**: Check the report to see where you can make improvements.

Doing this regularly while you develop and before deploying your site can help you identify large bundles earlier.

> ### Using Lighthouse for Performance Analysis

Reviewer 2: ```suggestion
### Using Lighthouse for performance analysis
```

> Google Lighthouse is a powerful tool for assessing Core Web Vitals, along with other important aspects such as performance, accessibility, best practices, and SEO. To get the most accurate results:

Reviewer 2: ```suggestion
[Google Lighthouse](https://developer.chrome.com/docs/lighthouse/overview) is a powerful tool for assessing Core Web Vitals, along with other important aspects such as performance, accessibility, best practices, and SEO. To get the most accurate results:
```

1. Use a production build of your Next.js application.
2. Run Lighthouse in incognito mode to prevent browser extensions from affecting the results.

> Lighthouse is available directly within Chrome DevTools, as well as a command-line tool, providing detailed reports on how your website performs against these vital metrics.

Reviewer 2: ```suggestion
Lighthouse is available directly within Chrome DevTools, as a command-line tool, and online through [Google PageSpeed Insights](https://pagespeed.web.dev).
```

> ### Utilizing Vercel's Dashboard (for Vercel Deployments)

Reviewer 2: ```suggestion
### Utilizing Vercel's dashboard (for Vercel deployments)
```

> If your Next.js application is deployed on [Vercel](https://vercel.com), you can leverage Vercel's built-in analytics dashboard to measure Core Web Vitals. This dashboard provides real-world performance insights based on actual visitor data, offering a practical view of how your website is performing in the field.

Reviewer 2: ```suggestion
If your Next.js application is deployed on [Vercel](https://vercel.com), you can leverage [Vercel's built-in analytics dashboard](https://vercel.com/analytics) to measure Core Web Vitals. This dashboard provides real-world performance insights based on actual visitor data, offering a practical view of how your website is performing in the field.
```

Once you are able to measure the loading performance, use the following strategies to improve it iteratively. That is, if you apply one strategy, measure the new performance and continue tweaking until you do not see much improvement. Then, you can move on to the next strategy.

- **Optimize Caching**: Use caching strategies as detailed in the caching section.
- **Server Actions Efficiency**: Optimize server actions for quick data retrieval and minimal processing.
- **Reduce JavaScript Bundle Sizes**: Remove unused JavaScript, and move data fetching and logic to the server.
  - Setting up your Code Editor to view import costs and sizes
  - Finding alternative smaller packages
  - Dynamically loading components and dependencies

> - **Streaming**: Leverage the Loading UI and Suspense to stream chunks in parallel for improved initial load.

Reviewer 2: Does it feel weird to list these here since they were already outlined and covered above? Should the "Loading performance" section be first? If we leave this here, should we add has links to the section above?

Author: I like the idea of moving the loading performance section up to be first

- Ensure your database and backend are deployed in the same region.
- Prevent sensitive data from being exposed to the client with [tainting](/docs/app/building-your-application/data-fetching/patterns#preventing-sensitive-data-from-being-exposed-to-the-client)
- Utilize logging for both the client and server.
- Ensure errors in nested route segments are handled gracefully with the [error.js file](/docs/app/building-your-application/routing/error-handling)
- Customize your 404 (Not Found) responses using the [`notFound()`](/docs/app/api-reference/functions/not-found) function and [`not-found.js` file](/docs/app/api-reference/file-conventions/not-found)
- Review [Supported Browsers and Features](/docs/architecture/supported-browsers)
- Optimize the user experience with built-in components:
  - [Images](/docs/app/building-your-application/optimizing/images)
  - [Fonts](/docs/app/building-your-application/optimizing/fonts)
  - [Scripts](/docs/app/building-your-application/optimizing/scripts)
  - [Link](/docs/app/building-your-application/routing/linking-and-navigating)

> - Follow the recommended [security best practices for Server Actions (/blog/security-nextjs-server-components-actions).

Reviewer: ```suggestion
- Follow the recommended [security best practices for Server Actions](/blog/security-nextjs-server-components-actions).
```

2. **Configure**: Add the bundle analyzer's settings to your `next.config.js`.

```js
const withBundleAnalyzer = require('@next/bundle-analyzer')({
  enabled: process.env.ANALYZE === 'true',
})
module.exports = withBundleAnalyzer({})
```

3. **Run Analysis**: Use `ANALYZE=true npm run build` to start looking at your file sizes.
4. **Review Findings**: Check the report to see where you can make improvements.

Doing this regularly while you develop and before deploying your site can help you identify large bundles earlier.

### Using Lighthouse for performance analysis

[Google Lighthouse](https://developer.chrome.com/docs/lighthouse/overview) is a powerful tool for assessing Core Web Vitals, along with other important aspects such as performance, accessibility, best practices, and SEO. To get the most accurate results:

1. Use a production build of your Next.js application.
2. Run Lighthouse in incognito mode to prevent browser extensions from affecting the results.

Lighthouse is available directly within Chrome DevTools, as a command-line tool, and online through [Google PageSpeed Insights](https://pagespeed.web.dev).

> ### Utilizing Vercel's dashboard (for Vercel deployments)

Reviewer 3: We might want to instead just link to the Vercel production checklist versus this section 👍

---
title: Going to Production
description: Before taking your Next.js application to production, here are some recommendations to ensure the best user experience.
---

Before taking your Next.js application to production, here are some recommendations to ensure the best user experience.

> ## Overview

Reviewer 4: nit, the overview doesn't really reflect the exact order of the item in the file

- Use [caching](/docs/app/building-your-application/caching) wherever possible.
- Ensure your database and backend are deployed in the same region.

> - Focus on leveraging [Server Components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) for efficient rendering. Server Components enable rendering on the server, which can significantly reduce the size of your client-side JavaScript bundle. Remember, Server Components can contain Client Components, offering flexibility in how you compose your application's UI. For an in-depth understanding of security best practices with Server Components, see our [blog post on security in Next.js Server Components and Actions](https://nextjs.org/blog/security-nextjs-server-components-actions).

Reviewer: Server Components and Server Actions feel like two separate concerns, maybe you can split this into bullet points: 

```suggestion
- Leverage [Server Components](https://nextjs.org/docs/app/building-your-application/rendering/server-components) as much as possible for improved performance. Unlike Client Components which require JavaScript to render on the client, Server Components render on the server, which significantly reduces the size of your client-side JavaScript bundle. See [Composition Patterns](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns) for best practices on using Server and Client Components. 
- Follow the recommended [security best practices for Server Actions](https://nextjs.org/blog/security-nextjs-server-components-actions).
```

Also added link to relevant docs.

- Prevent sensitive data from being exposed to the client with [tainting](/docs/app/building-your-application/data-fetching/patterns#preventing-sensitive-data-from-being-exposed-to-the-client)
- Ensure errors in nested route segments are handled gracefully with the [error.js file](https://nextjs.org/docs/app/building-your-application/routing/error-handling)
- Customize your 404 (Not Found) responses using the [`notFound()`](/docs/app/api-reference/functions/not-found) function and [`not-found.js` file](https://nextjs.org/docs/app/api-reference/file-conventions/not-found)

> - Ensure you are [measuring performance](/docs/app/building-your-application/optimizing/analytics).

Reviewer 4: This point could come after the following one, e.g.
- use lighthouse to test your product's performance
- on production, collect real user metrics (RUM) with analytics

- Run [Lighthouse](https://developers.google.com/web/tools/lighthouse) to check for performance, best practices, accessibility, and SEO. For best results, use a production build of Next.js and use incognito in your browser so results aren't affected by browser extensions.
- Review [Supported Browsers and Features](/docs/architecture/supported-browsers).
- Optimize the user experience with built-in components:
  - [Images](/docs/app/building-your-application/optimizing/images)
  - [Fonts](/docs/app/building-your-application/optimizing/fonts)
  - [Scripts](/docs/app/building-your-application/optimizing/scripts)
  - [Link](/docs/app/building-your-application/routing/linking-and-navigating)
- Improve [loading performance](#loading-performance)
- Consider adding a [Content Security Policy](/docs/app/building-your-application/configuring/content-security-policy)

## Caching

> Caching is essential for improving response times and reducing requests to backend services. The App Router in Next.js offers a robust [caching mechanism](https://nextjs.org/docs/app/building-your-application/caching) that supports static and dynamic rendering, as well as [revalidation](/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#revalidating-data).

Reviewer 4: nit

> that supports static and dynamic rendering

maybe 
>  that allows to leverage static and dynamic rendering

### Static assets

Next.js automatically adds caching headers to immutable assets in **`/public`** or other static directories. Assets like CSS, static images, and other media are ideal candidates for long-term caching as they rarely change. The default caching headers applied are:

```jsx
Cache-Control: public, max-age=31536000, immutable
```

> This setup ensures that these assets are cached efficiently, reducing load times for returning users.

Reviewer 4: are cached efficiently in the browser

### Data caching

Next.js's built-in [Data Cache](/docs/app/building-your-application/caching#data-cache) persists results of data fetches across server requests and deployments. By default, fetch requests are cached, allowing data to be reused in subsequent requests. This feature is particularly useful for content that doesn't change frequently.

> ### Revalidation

Reviewer 4: should this just be a sub item of the item above?

The caching behavior of data requests can be controlled through various strategies. Next.js supports [time-based revalidation](/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#time-based-revalidation) (automatically refreshing the cache after a set period) and [on-demand revalidation](/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#on-demand-revalidation) (refreshing the cache based on specific events).

> ### Caching in Server Actions

Reviewer 4: maybe

> Controlling the cache/caching with Server Actions

While Server Actions in Next.js are primarily used for mutations (like POST, PUT, DELETE operations), they also play a crucial role in cache management. After a mutation, it's crucial to ensure that the cached data reflects the latest state. Next.js facilitates this by providing APIs such as [`revalidatePath`](https://nextjs.org/docs/app/api-reference/functions/revalidatePath) and [`revalidateTag`](https://nextjs.org/docs/app/api-reference/functions/revalidateTag). These APIs can be used to purge or revalidate the cache immediately after a mutation. This capability ensures that any changes made by Server Actions are promptly reflected in the cached data, maintaining data consistency and accuracy.

### Opting Out of Caching

There are scenarios where caching might not be desirable as users need the most up-to-date content. Next.js allows [opting out of the Data Cache](/docs/app/building-your-application/caching#opting-out-1) for specific fetch requests. This can be achieved with settings such as `cache: 'no-store'` or `revalidate: 0`.

> ## Reducing JavaScript bundle sizes

Reviewer 4: this kinda belongs as an item of the loading performance item

How to use `@next/bundle-analyzer`:

1. **Install**: Run `npm install @next/bundle-analyzer webpack-bundle-analyzer`.
2. **Configure**: Add the bundle analyzer's settings to your `next.config.js`.

```js
const withBundleAnalyzer = require('@next/bundle-analyzer')({
  enabled: process.env.ANALYZE === 'true',
})
module.exports = withBundleAnalyzer({})
```

3. **Run Analysis**: Use `ANALYZE=true npm run build` to start looking at your file sizes.
4. **Review Findings**: Check the report to see where you can make improvements.

Doing this regularly while you develop and before deploying your site can help you identify large bundles earlier.

## Logging

Since Next.js runs on both the client and server, there are multiple forms of logging supported:

- `console.log`: Works in both browser and server. In the browser, it logs to the console. On the server, it outputs to the terminal.

> - `process.stdout.write`: Used on the server for more control over the output format, writing directly to stdout.

Reviewer 4: Do we want to recommend that? I don't see it really often

If you want a structured logging package, we recommend [Pino](https://www.npmjs.com/package/pino). If you're using Vercel, there are [pre-built logging integrations](https://vercel.com/integrations?utm_source=next-site&utm_medium=docs&utm_campaign=next-website#logging) compatible with Next.js.

## Error Handling

> Proper error handling is key to keeping your application stable and providing a good user experience. This is especially important with the Dynamic Route Segments and Server Actions in the App Router.

Reviewer 4: > This is especially important with the Dynamic Route Segments

why?

---
title: Going to Production
description: Before taking your Next.js application to production, here are some recommendations to ensure the best user experience.
---

Before taking your Next.js application to production, here are some recommendations to ensure the best user experience.

> ## Overview

Reviewer 5: This blog should be mentioned somewhere: https://nextjs.org/blog/security-nextjs-server-components-actions

- Use [caching](/docs/app/building-your-application/caching) wherever possible.
- Ensure your database and backend are deployed in the same region.

> - Aim to reduce client-side JavaScript bundle sizes by [moving the "use client" boundaries down the React component tree](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#moving-client-components-down-the-tree).

Reviewer 5: We don't have to recommend moving the use client boundaries down, instead it should explain leveraging Server Components. It's not exactly that the tree stops when there is a client component, they can wrap server components content.

Reviewer: I think we're saying the same thing here but with different words. One mistake I often see people doing is opting into Client Components higher up the tree and/or passing large data objects to the client, thus increasing their JavaScript bundle sizes. 

Hence why it might be good to be explicit about "moving the boundaries down the tree" or in other words moving client components down the leaves, and linking to the docs section that recommends this pattern.

Since Server Components are the default, it's more about asking people to be mindful of when/where they place the client boundaries.

Reviewer: I'll write a suggestion below 🙏🏼

- Prevent sensitive data from being exposed to the client with [tainting](/docs/app/building-your-application/data-fetching/patterns#preventing-sensitive-data-from-being-exposed-to-the-client)
- Ensure errors in nested route segments are handled gracefully with the [error.js file](https://nextjs.org/docs/app/building-your-application/routing/error-handling)
- Customize your 404 (Not Found) responses using the [`notfound()`](/docs/app/api-reference/functions/not-found) function and [`not-found.js` file](https://nextjs.org/docs/app/api-reference/file-conventions/not-found)
- Ensure you are [measuring performance](/docs/app/building-your-application/optimizing/analytics).

> - Run [Lighthouse](https://developers.google.com/web/tools/lighthouse) to check for performance, best practices, accessibility, and SEO. For best results, use a production build of Next.js and use incognito in your browser so results aren't affected by extensions.

Reviewer 5: ```suggestion
- Run [Lighthouse](https://developers.google.com/web/tools/lighthouse) to check for performance, best practices, accessibility, and SEO. For best results, use a production build of Next.js and use incognito in your browser so results aren't affected by browser extensions.
```

```jsx
Cache-Control: public, max-age=31536000, immutable
```

This setup ensures that these assets are cached efficiently, reducing load times for returning users.

### Data caching

Next.js's built-in [Data Cache](/docs/app/building-your-application/caching#data-cache) persists results of data fetches across server requests and deployments. By default, fetch requests are cached, allowing data to be reused in subsequent requests. This feature is particularly useful for content that doesn't change frequently.

### Revalidation

The caching behavior of data requests can be controlled through various strategies. Next.js supports [time-based revalidation](/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#time-based-revalidation) (automatically refreshing the cache after a set period) and [on-demand revalidation](/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#on-demand-revalidation) (refreshing the cache based on specific events).

### Caching in Server Actions

> While Server Actions in Next.js are primarily used for mutations (like POST, PUT, DELETE operations), they also play a crucial role in cache management. After a mutation, it's crucial to ensure that the cached data reflects the latest state. Next.js facilitates this by providing APIs such as `revalidatePath` and `revalidateTag`. These APIs can be used to purge or revalidate the cache immediately after a mutation. This capability ensures that any changes made by Server Actions are promptly reflected in the cached data, maintaining data consistency and accuracy.

Reviewer 5: Seems to be missing links to revalidatePath / revalidateTag.

### Opting Out of Caching

There are scenarios where caching might not be desirable as users need the most up-to-date content. Next.js allows [opting out of the Data Cache](/docs/app/building-your-application/caching#opting-out-1) for specific fetch requests. This can be achieved with settings such as `cache: 'no-store'` or `revalidate: 0`.

## Reducing JavaScript bundle sizes

> Making sure your website sends less JavaScript to the browser is important for speed. To help you understand and manage your JavaScript bundles, you can use these tools to visualize your bundles and/or identify large dependencies.

Reviewer 5: > Making sure your website sends less JavaScript to the browser is important for speed

You might want to change the wording of this because JS size doesn't specifically make Core Web Vitals faster in all cases, it's not a silver bullet basically.

- [Import Cost](https://marketplace.visualstudio.com/items?itemName=wix.vscode-import-cost) – Shows the size of each package you import right in VSCode.
- [Package Phobia](https://packagephobia.com/) – Helps you see the size impact of adding new dev dependencies to your project.
- [Bundle Phobia](https://bundlephobia.com/) - Tells you how much a new dependency could increase your JavaScript size.
- [Webpack Bundle Analyzer](https://github.com/vercel/next.js/tree/canary/packages/next-bundle-analyzer) – Creates a visual, interactive map of your webpack output files to show where space is used.
- [bundlejs](https://bundlejs.com/) - A browser tool for bundling and minifying projects, showing compressed sizes.

Additionally, make sure to implement the recommended optimization strategies.

> - Optimize Server Actions - Focus on simplifying and speeding up your server actions. This can be achieved by optimizing database queries, reducing external API calls, and ensuring server-side computations are efficient. Efficient server-side processing helps reduce the time it takes for your pages to load and improves the overall performance of your website.

Reviewer 5: This one feels off under the `reduce JS size` heading

- Minimize Client-Side JavaScript - It's important to keep JavaScript on the client-side as small as possible for quicker website loading and a better user experience.

> - Moving client components down the tree or carefully considering where to place the "use client" boundaries. In other words, moving more rendering / logic to the server using Server Components.

Reviewer 5: Similar to https://github.com/vercel/next.js/pull/59304/files#r1419012086

If an error is thrown inside a Server Component, Next.js forwards an `Error` object to the nearest `error.js` file. This object includes a generic message and a digest property in production as a security measure to prevent leaking sensitive details.

## Loading Performance

To effectively improve loading performance, it's essential to measure the user experience using [Core Web Vitals](https://vercel.com/blog/core-web-vitals?utm_source=next-site&utm_medium=docs&utm_campaign=next-website). These are a set of metrics developed by Google to gauge the quality of user experience on web pages, influencing Google's search rankings.

### Using Lighthouse for Performance Analysis

Google Lighthouse is a powerful tool for assessing Core Web Vitals, along with other important aspects such as performance, accessibility, best practices, and SEO. To get the most accurate results:

1. Use a production build of your Next.js application.
2. Run Lighthouse in incognito mode to prevent browser extensions from affecting the results.

> Lighthouse can be used both as a browser extension and a command-line tool, providing detailed reports on how your website performs against these vital metrics.

Reviewer 5: Lighthouse is built-in to the Chrome devtools so yo don't have to use a browser extension 👍
